---
title: "Exploratory analysis of a GEO dataset"
author: "Ferriol Calvet"
date: "May 15, 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction
This exercise and all the code will be available in the following link:
https://github.com/FerriolCalvet/Calvet_Ferriol-OmicsTechniques

## Summary
The following description was given: *Nebulin is a giant filamentous protein that is coextensive with the actin filaments of the skeletal muscle sarcomere. Nebulin mutations are the main cause of nemaline myopathy (NEM), with typical NEM adult patients having low expression of nebulin, yet the roles of nebulin in adult muscle remain poorly understood. To establish nebulinâ€™s functional roles in adult muscle we performed studies on a novel conditional nebulin KO (Neb cKO) mouse model in which nebulin deletion was driven by the muscle creatine kinase (MCK) promotor. Neb cKO mice are born with high nebulin levels in their skeletal muscle but within weeks after birth nebulin expression rapidly falls to barely detectable levels Surprisingly, a large fraction of the mice survives to adulthood with low nebulin levels (<5% of control), contain nemaline rods, and undergo fiber-type switching towards oxidative types. These microarrays investigate the changes in gene expression when nebulin is deficient.*

## Experimental design
Two skeletal muscle groups were studied: Quadriceps (which is markedly smaller in the Neb cKO mice relative to control) and Soleus (which is not significantly smaller in the Neb cKO relative to control). Six biological replicates for each muscle group were selected; all are age-matched males.

## Citation(s)
This data was published in: 
*Li F, Buck D, De Winter J, Kolb J et al. Nebulin deficiency in adult muscle causes sarcomere defects and muscle-type-dependent changes in trophicity: novel insights in nemaline myopathy. Hum Mol Genet 2015 Sep 15;24(18):5219-33. PMID: 26123491*


## Data description
We have two files:
    Expression matrix, saved as _expres2_
    Targets' information, saved as _targets\_info_ 

```{r}
expres2 <- read.delim("expres2.txt", row.names=1)
targets_info <- read.table("targets.txt", header = T)
targets_info$group <- c(rep("1",6), rep("2",6), rep("3",6), rep("4",6))
# View(expres2)
show(targets_info)
```

The expression matrix contains the different 24 samples in the columns. These samples can be grouped into 4 different groups, with 6 replicates per each group. Note that in the output of the previous chunk you can see these different groups and replicates.
Then for each of the samples we can find 35557 observations, this means that we have information for the expression of 35557 different transcripts.


### Change sample names to informative names
This is optional, we can keep working with the original names
```{r}
# new_names <- paste(targets_info$muscle, targets_info$genotype, targets_info$replicate)

# colnames(expres2) <- new_names
# colnames(expres2)

# head(expres2)
```

## Analysis: Observation and summary


```{r}
summary(expres2)
```
If we look at the summary results, we see that the values vary a lot among the same sample. 
We can see that the difference between the minimum and the maximum is of the order of magnitude 4, being in most of the cases from around 2 until around 10000.  
Then the mean never falls inside the interquantile range, as this measurement is highly affected by extreme values, maybe we should correct this at some point. In order to have a clearer view of the distribution of our data, we will now make some boxplots.

```{r}
boxplot(expres2)
# we apply log transformation because it is a possible way to correct the data
boxplot(log(expres2))

expres2 <- log(expres2)
```
When we make the first boxplot, we can see check that yes, the data is very dispersed so we cannot appreciate any box, a possible transformation that can be applied to our data is the logarithmic transformation. If we compare the two boxplots, we can see that after the transformation the information of the different groups can be compared easily and with more balanced values.

##### Now start with the real analysis
After applying the logarithm transformation to our, we can follow with the analysis.
```{r}
hist(expres2[,1])

row_means_apply <- apply(expres2,1, mean)
hist(row_means_apply)
```



```{r}
expres2 <- read.delim("expres2.txt", row.names=1)
pcs <- prcomp(expres2)
names(pcs)
barplot(pcs$sdev)
plot(pcs$rotation[,1], pcs$rotation[,2], col=targets_info$group, main="Representation of first two principal components")
```

# ExpressionSet construction 1

```{r, include=FALSE}
require(Biobase)
```

```{r}
targets_column_names <- data.frame(labelDescription= colnames(targets_info))

## ----simulateInfo--------------------------------------------------------
myInfo=list(myName="Ferriol Calvet", myLab="Bioinformatics Lab", 
            myContact="ferriol.calvet@alum.esci.upf.edu", myTitle="Practical Exercise on ExpressionSets")
show(myInfo)


## ----creaExpressionSet1--------------------------------------------------
define_matrix <- as.matrix(expres2)
myEset <- ExpressionSet(define_matrix)
class(myEset)
show(myEset)


## ----AnnotatedDataFrame2-------------------------------------------------
columnDesc <-  data.frame(labelDescription= colnames(targets_info))
rownames(targets_info) <- targets_info$sampleName

myAnnotDF <- new("AnnotatedDataFrame", data=targets_info, varMetadata= columnDesc)
show(myAnnotDF)

# Version 1
phenoData(myEset) <- myAnnotDF
show(myEset)

# Version 2
myEset <- ExpressionSet(assayData=define_matrix, phenoData=myAnnotDF)
show(myEset)

## -----add-features-names----------------------------------------------------------------
myGenes <- rownames(expres2)
myEset <- ExpressionSet(assayData=define_matrix, 
                        phenoData=myAnnotDF, 
                        featureNames =myGenes)

myInfo=list(myName="Ferriol Calvet", myLab="Bioinformatics Lab", 
            myContact="ferriol.calvet@alum.esci.upf.edu", myTitle="Practical Exercise on ExpressionSets")

## ----label=MIAME---------------------------------------------------------
myDesc <- new("MIAME", name= myInfo[["myName"]],
              lab= myInfo[["myLab"]],
              contact= myInfo[["myContact"]] ,
              title=myInfo[["myTitle"]])

## ------------------------------------------------------------------------
myEset <- ExpressionSet(assayData=define_matrix, 
                        phenoData=myAnnotDF,
                        fetureNames =myGenes,
                        experimentData = myDesc)
```

# Analyze this ExpressionSet built manually

```{r}
dim(exprs(myEset))
class(phenoData(myEset))

# This two instructions are the same
class(pData(phenoData(myEset)))
head(pData(phenoData(myEset)))
class(pData(myEset))
head(pData(myEset))


## ------------------------------------------------------------------------
smallEset <- myEset[1:50,13:24]

dim(exprs(smallEset))
dim(pData(smallEset))
head(pData(smallEset))
all(colnames(exprs(smallEset))==rownames(pData(smallEset)))
```
For example we can select a small dataset of only two groups, in this case both selected groups are the ones with expression data of the soleus muscle. And then in the amount of genes, we have only selected the first 50 genes.

# ExpressionSet construction 2

```{r, include=FALSE}
require(GEOquery)
```

```{r}
gse <- getGEO("GSE70213")
class(gse)
names(gse)
gse[[1]]
esetFromGEO <- gse[[1]]

```

# Analyze this ExpressionSet automatically generated
```{r}
## ------------------------------------------------------------------------
dim(exprs(myEset))
class(phenoData(myEset))

# This two instructions are the same
class(pData(phenoData(myEset)))
head(pData(phenoData(myEset)))
class(pData(myEset))
head(pData(myEset))


## ------------------------------------------------------------------------
smallEset <- myEset[1:50,13:24]

dim(exprs(smallEset))
dim(pData(smallEset))
head(pData(smallEset))
all(colnames(exprs(smallEset))==rownames(pData(smallEset)))
```
We can see that now the results are the same as before.